lib/compilers/rust.js

```diff
@@ -53,6 +67,40 @@ export class RustCompiler extends BaseCompiler {
         return [];
     }
 
+    // Currently called for getting macro expansion and HIR.
+    // It returns the content of the output file created after using -Z unpretty=<unprettyOpt>.
+    // The outputFriendlyName is a free form string used in case of error.
+    async generateRustUnprettyOutput(inputFilename, options, unprettyOpt, outputFilename, outputFriendlyName){
+        const execOptions = this.getDefaultExecOptions();
+
+        const rustcOptions = [...options];
+        rustcOptions.splice(options.indexOf('-o', 2));
+        rustcOptions.push(inputFilename, '-o', outputFilename, `-Zunpretty=${unprettyOpt}`);
+
+        const output = await this.runCompiler(this.compiler.exe, rustcOptions, inputFilename,
+                                              execOptions);
+        if (output.code !== 0) {
+            return [{text: `Failed to run compiler to get Rust ${outputFriendlyName}`}];
+        }
+        if (await fs.exists(outputFilename)) {
+            const content = await fs.readFile(outputFilename, 'utf-8');
+            return content.split('\n').map((line) => ({
+                text: line,
+            }));
+        }
+        return [{text: 'Internal error; unable to open output path'}];
+    }
+
+    async generateRustMacroExpansion(inputFilename, options) {
+        const macroExpPath = this.getRustMacroExpansionOutputFilename(inputFilename);
+        return this.generateRustUnprettyOutput(inputFilename, options, 'expanded', macroExpPath, 'Macro Expansion');
+    }
+
+    async generateRustHir(inputFilename, options) {
```

Comment:
Removing the definitions in BaseCompiler and moving them here while still referencing `this.generateRustHir/MacroExpansion` isn't really sound. While there shouldn't be any requests taking that route, there is nothing preventing the BaseCompiler from blowing up here.

@mattgodbolt and I have talked about this previously, and I believe we concluded that we should keep them in BaseCompiler until we can come up with a better solution.

Reply:
Yes, wasn't really happy with the design here. I decided to move it anyway so rust specifics stays in rust.js as much as possible. But the base class is really doing more than the shared behavior and is already full of compiler specifics.
Maybe refactoring the `doCompilation` and move the several specific invocations below to the sub-classes could help.

```
        const [
            asmResult,
            astResult,
            irResult,
            rustHirResult,
            rustMacroExpResult,
            toolsResult,
        ] = await Promise.all([
            this.runCompiler(this.compiler.exe, options, inputFilenameSafe, execOptions),
            (makeAst ? this.generateAST(inputFilename, options) : ''),
            (makeIr ? this.generateIR(inputFilename, options, filters) : ''),
            (makeRustHir ? this.generateRustHir(inputFilename, options) : ''),
            (makeRustMacroExp ? this.generateRustMacroExpansion(inputFilename, options) : ''),
            Promise.all(this.runToolsOfType(tools, 'independent', this.getCompilationInfo(key, {
                inputFilename,
                dirPath,
                outputFilename,
            }))),
        ]);
```

I'll move them back.

Reply:
> Generally looks good, just a bit unsure about the BaseCompiler change I've mentioned in the comment above.
> 
> I also think we should find an even nicer way to share functionality across panes (as IR, Rust MIR, Rust Macro and Rust HIR are essentially the same pane passing different API parameters) but that's out of scope for this PR. Thanks!

Adding a new pane really feels like useless (and a little bit of shame) duplication, but I really don't know anything about js and the frameworks to even have a remote idea of how to make this better, sorry.